district_data_import.py

```diff
@@ -0,0 +1,415 @@
+                         'enroll_status', 'Comment', 'PTA_BCE_Permit']
+
+SCHOOL_YEAR = 'SY2016-17'
+TODAY_STR = datetime.date.today().strftime('%m/%d/%Y')
+
+FILENAME_STUDENT_UPDATES = '01-student-updates.csv'
+FILENAME_NEWSTUDENT = '02-new-students.csv'
+FILENAME_NEWDONOR = '03-new-donors.csv'
+FILENAME_DONOR_UPDATES = '04-donor-updates.csv'
+FILENAME_DONOR_UPDATE_MESSAGES = '05-donor-manual-updates.txt'
+
+last_seq_values = {}
+def gen_next_value(seq):
+    next = last_seq_values.get(seq, 0) + 1
+    last_seq_values[seq] = next
+    return next
+
+def gen_donor_id():
+    return -1 * gen_next_value('DONOR_ID')
+
+def gen_other_id():
+    return -1 * gen_next_value('OTHER_ID')
+
+if len(sys.argv) != 3:
+    usage()
+
+filename_dp_report_271 = sys.argv[1]
+filename_district_data = sys.argv[2]
+
+print("Input files:")
+
+# DP donors keyed by DONOR_ID
+dp_unmodified_donorrecords = {}
+dp_donorrecords = {}
+
+# DP students keyed by OTHER_ID
+dp_unmodified_studentrecords = {}
+dp_studentrecords = {}
+
+# Multi-maps for various ids
+dp_donor_id_to_other_ids = defaultdict(list)
+dp_stu_number_to_other_ids = defaultdict(list)
+
+# Load DP data
+includes_nomail = False
+for row in utils.load_csv_file(filename_dp_report_271, DP_REPORT_271_HEADERS):
+    # Process donor-level info
+    donor_id = row['DONOR_ID']
+    dp_donorrecord = dict((header, row[header]) for header in DP_REPORT_271_DONOR_HEADERS)
+    if donor_id in dp_donorrecords:
+        if dp_donorrecord != dp_unmodified_donorrecords[donor_id]:
+            raise ValueError("Unexpected differences in donors. Assumptions must be incorrect. Expected: %s, Actual: %s" %
+                             (dp_unmodified_donorrecords[donor_id], dp_donorrecord))
+    else:
+        dp_unmodified_donorrecords[donor_id] = dp_donorrecord.copy()
+        dp_donorrecords[donor_id] = dp_donorrecord
+
+    # Process student-level info
+    other_id = row['OTHER_ID']
+    if other_id in dp_studentrecords:
+        raise ValueError("Found a duplicate OTHER_ID in report 271, the report's assumptions are now violated")
+    dp_studentrecord = dict((header, row[header]) for header in DP_REPORT_271_STUDENT_HEADERS)
+    dp_unmodified_studentrecords[other_id] = dp_studentrecord.copy()
+    dp_studentrecords[other_id] = dp_studentrecord
+
+    # Compute many-to-many relationships
+    dp_donor_id_to_other_ids[donor_id].append(other_id)
+    if row['STU_NUMBER']:
+        dp_stu_number_to_other_ids[row['STU_NUMBER']].append(other_id)
+
+    if row['NOMAIL'] == 'Y':
+        includes_nomail = True
+
+if not includes_nomail:
+    print("%s must include \"NO MAIL\" donors. Please select 'Include \"NO MAIL\" Names' and regenerate the report." % filename_dp_report_271)
+    sys.exit(1)
+
+# Load district data keyed off student number ("system id" there)
+district_records = {}
+for row in utils.load_csv_file(filename_district_data, DISTRICT_DATA_HEADERS):
+    district_records[row['SystemID']] = row
+
```

Comment:
You understand correctly: 1) Import csv file into memory, 2) Overlay district data on in-memory structure, 3) Output CSV files to bring DP into line with that in-memory structure.

I think you're suggesting that we should not do step 2 but should instead have a series of side-effect-free 2/3 combinations? If that's the case, I don't think so. That's essentially what we had before, and I think it was very limiting. For example, setting the NOMAIL stuff properly requires that you have an up-to-date version of everything in memory in order to count students. Also, this allows us to do sophisticated stuff like filter the fields eligible for modification simply by tweaking step 3. This is very useful for testing.

I think I see how to extract all the DP data to a separate class, which I'll go ahead and do. I'll let you know once I've updated the PR.